### PR TITLE
Retry the submit-application request to multiple nodes

### LINKED ISSUE #67: Regression introduced by NodePort PR
Just debugged with @mccheah , and we found an issue with the NodePort implementation of the jar upload process (https://github.com/apache-spark-on-k8s/spark/pull/30)

The code [assumes that the port on all returned k8s nodes is accessible](https://github.com/apache-spark-on-k8s/spark/blob/a4abee3126f308e8c3e01c47ab7f8251bc5c1cbe/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala#L545), but we found in testing with my `kubeadm`-created cluster that actually the k8s master does NOT forward the traffic to the pod behind the service.  Only the non-master k8s nodes do.

Setup:

myhost-16 is an apiserver, and myhost-17/18/19 are non-masters but part of the k8s cluster.  Cluster assembled using https://kubernetes.io/docs/getting-started-guides/kubeadm/

response returned from jetty (the `KubernetesSparkRestApi`):
`curl -XGET http://myhost-17.mydomain:30827/v1/submissions/ping`
`curl -XGET http://myhost-18.mydomain:30827/v1/submissions/ping`
`curl -XGET http://myhost-19.mydomain:30827/v1/submissions/ping`

request times out:
`curl -XGET http://myhost-16.mydomain:30827/v1/submissions/ping`

Proposed solution: fail over in the liveness pings from non-responsive nodes to other nodes.  This fixes the problem with the master, and more generally also nodes failing out of the cluster.

### PR DESCRIPTION
**Pull request by mccheah**
Closes #67. Use a custom Feign Target which can try making requests against multiple servers.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes
+
+import feign.{Request, RequestTemplate, RetryableException, Retryer, Target}
+import scala.reflect.ClassTag
+import scala.util.Random
+
+private[kubernetes] class MultiServerFeignTarget[T : ClassTag](
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala:23]: This is inspired by [http-remoting](https://github.com/palantir/http-remoting/blob/develop/jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/feignimpl/FailoverFeignTarget.java).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 2]
```diff
@@ -45,13 +45,24 @@ private[spark] object HttpClientUtil {
       .registerModule(new DefaultScalaModule)
       .setDateFormat(JacksonMessageWriter.makeISODateFormat)
     objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
-    val clazz = implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]]
+    val target = new MultiServerFeignTarget[T](uris)
+    val baseHttpClient = new feign.okhttp.OkHttpClient(httpClientBuilder.build())
+    val resetTargetHttpClient = new Client {
+      override def execute(request: Request, options: Options): Response = {
+        val response = baseHttpClient.execute(request, options)
+        if (response.status() >= 200 && response.status() < 300) {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala:53]: do you mean 2xx responses to reset that target?  I'd expect non-2xx responses to be what triggers a reset and fails over to another URI in the list

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  override def `type`(): Class[T] = {
+    implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]]
+  }
+
+  override def url(): String = threadLocalShuffledServers.get.head
+
+  /**
+    * Cloning the target is done on every request, for use on the current
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala]: scalastyle is complaining about this -- the Spark project uses javadoc style comments not scaladoc

**Review by ash211 — COMMENTED** [review]
I'll give this a shot and see if it fixes the issue in my test environment.

**Comment by ash211** [comment]
Hmm, so this works but the failover takes a significant amount of time to fail over:

With this patch:
```
2017-02-01 04:09:22 INFO  ServerConnector:266 - Started ServerConnector@769a1df5{HTTP/1.1}{org-apache-spark-examples-sparkpi-1485922157696:7077}
2017-02-01 04:09:22 INFO  Server:379 - Started @1294ms
2017-02-01 04:09:22 INFO  Utils:54 - Successfully started service on port 7077.
2017-02-01 04:09:22 INFO  KubernetesSparkRestServer:54 - Started REST server for submitting applications on port 7077
2017-02-01 04:10:05 INFO  SparkContext:54 - Running Spark version 2.2.0-SNAPSHOT
2017-02-01 04:10:05 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2017-02-01 04:10:05 INFO  SparkContext:54 - Submitted application: Spark Pi
2017-02-01 04:10:05 INFO  SecurityManager:54 - Changing view acls to: root
2017-02-01 04:10:05 INFO  SecurityManager:54 - Changing modify acls to: root
2017-02-01 04:10:05 INFO  SecurityManager:54 - Changing view acls groups to:
```
(43sec gap)

```
2017-02-01 06:05:47 INFO  Server:379 - Started @1358ms
2017-02-01 06:05:47 INFO  Utils:54 - Successfully started service on port 7077.
2017-02-01 06:05:47 INFO  KubernetesSparkRestServer:54 - Started REST server for submitting applications on port 7077
2017-02-01 06:06:08 INFO  SparkContext:54 - Running Spark version 2.2.0-SNAPSHOT
2017-02-01 06:06:08 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2017-02-01 06:06:08 INFO  SparkContext:54 - Submitted application: Spark Pi
2017-02-01 06:06:08 INFO  SecurityManager:54 - Changing view acls to: root
```

(21sec gap)

```
2017-02-01 06:07:39 INFO  Server:379 - Started @1406ms
2017-02-01 06:07:39 INFO  Utils:54 - Successfully started service on port 7077.
2017-02-01 06:07:39 INFO  KubernetesSparkRestServer:54 - Started REST server for submitting applications on port 7077
2017-02-01 06:08:20 INFO  SparkContext:54 - Running Spark version 2.2.0-SNAPSHOT
2017-02-01 06:08:20 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2017-02-01 06:08:20 INFO  SparkContext:54 - Submitted application: Spark Pi
```

(41sec gap)


With a separate patch that just drops the master:
```
diff --git a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala
index fed9334dbb..fa1df14666 100644
--- a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala
@@ -541,7 +541,7 @@ private[spark] class Client(
       .getNodePort
     // NodePort is exposed on every node, so just pick one of them.
     // TODO be resilient to node failures and try all of them
-    val node = kubernetesClient.nodes.list.getItems.asScala.head
+    val node = kubernetesClient.nodes.list.getItems.asScala.tail.head
     val nodeAddress = node.getStatus.getAddresses.asScala.head.getAddress
     val urlScheme = if (driverSubmitSslOptions.enabled) {
       "https"
```

logs:
```
2017-01-31 22:27:12 INFO  Server:379 - Started @1332ms
2017-01-31 22:27:12 INFO  Utils:54 - Successfully started service on port 7077.
2017-01-31 22:27:12 INFO  KubernetesSparkRestServer:54 - Started REST server for submitting applications on port 7077
2017-01-31 22:27:19 INFO  SparkContext:54 - Running Spark version 2.2.0-SNAPSHOT
2017-01-31 22:27:19 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2017-01-31 22:27:19 INFO  SparkContext:54 - Submitted application: Spark Pi
2017-01-31 22:27:19 INFO  SecurityManager:54 - Changing view acls to: root
```

(7sec gap)


It seems like it takes a significant amount of time to fail over to a different endpoint.  Does it make sense to reduce the timeout for the pings when checking for liveness so it fails over faster?  Is there something else we can do to reduce the time spent waiting for failover?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala (hunk 2 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala:53]: See [remoting](https://github.com/palantir/http-remoting/blob/develop/jaxrs-clients/src/main/java/com/palantir/remoting1/jaxrs/feignimpl/FailoverFeignTarget.java#L151) - not sure, what's the right thing to check here though?

**Comment by foxish** [comment]
Is there a reason we don't combine the two? It seems like we should drop the master from the candidate list in any case before trying to connect to the nodes.

**Comment by foxish** [comment]
Just picking a single node at random from among the rest should also suffice. Kube-proxy is a critical system pod and *should* be running on all non-master nodes. If it's not running, there are usually issues on that node, and the administrator would need to step in.

**Comment by mccheah** [comment]
Hm, I think the original implementation this was inspired from assumes that the servers it failed to connect to may become available later. This might not be the same semantics as to what we're trying to accomplish, but as a principle it would be neater to not make the `MultiServerFeignTarget` opinionated towards the particular use case of the HTTP client.

**Comment by ash211** [comment]
Dropping the master from the node list either way seems like a good idea -- it would reduce the time for failover at least in my case, and it sounds like kube proxy on the master node isn't guaranteed anyway.

@mccheah what do you think of dropping the master node from the list of URIs?

**Comment by mccheah** [comment]
If we can filter the nodes at the level above the `MultiServerFeignTarget` then that would be better than filtering when we see a failed request at the HTTP level. However, this breaks local setup with e.g. `minikube` since in local mode the master also runs the proxy. It would be better to be able to tell if any given node is not running the proxy - @foxish is there any way to get that more precisely than just filter out the master node?

**Comment by foxish** [comment]
We can use the Node's `.spec.unschedulable=true|false` to decide whether kube-proxy is running on it. This seems to be the preferred way of finding nodes which we can hit. The networking sig is working on a different way to expose kube-proxy status via the node but it hasn't been merged yet. Our best bet I think would be to use the `unschedulable` field. Typically master nodes will have it set to `true`. In case of minikube, the master's `.spec.unschedulable` should not be marked `true`.

**Comment by foxish** [comment]
Edited my above comment: fixed, as I had accidentally inverted the logic.

**Comment by ash211** [comment]
my `kubeadm`-created cluster I'm not seeing a `.spec.unschedulable` field on any of the nodes (master or worker).  This is with k8s 1.5.1.

`$ kubectl get node <HOST> -o yaml | grep -i unschedul`

What normally creates that on a node's spec?

I do see a taint on the node -- maybe that could be used as a proxy for whether the nodeport is running on a node?

```
metadata:
  annotations:
    scheduler.alpha.kubernetes.io/taints: '[{"key":"dedicated","value":"master","effect":"NoSchedule"}]'
...
```

**Comment by foxish** [comment]
The node controller is responsible for setting `.spec.unschedulable`, and I can see that it is set on my 1.5.2 cluster. It is interesting that your master node has only the taint set. The taint has the same effect, but is the new method of marking nodes as unschedulable. I'm not sure what happened with the kubeadm bootstrapping process here and I'll file an issue with them. Could you show me the output of `kubectl get nodes` to see if the master has been marked as `SchedulingDisabled`? 

The safest thing would be to check the taints annotation, **and** the `.spec.unschedulable` fields. Both of them have exactly the same effect if `Unschedulable` is set. We will need to update it in future as the annotation evolves to beta, and then becomes a proper API field. :(

**Comment by mccheah** [comment]
I added more logging to the `Retry` code as I think some errors are being swallowed up there also.

@foxish - what's the exact annotation key and value I need to check? `node.getMetadata.getAnnotations` returns a `java.util.Map<String, String>` - presumably I need to find the specific key-value pair to filter out nodes by here.

**Comment by ash211** [comment]
@foxish as you requested (hostnames redacted):

```
[user@host-766416 ~]$ kubectl get nodes
NAME                 STATUS         AGE
host-766416.domain   Ready,master   8d
host-766417.domain   Ready          8d
host-766418.domain   Ready          8d
host-766419.domain   Ready          8d
[user@host-766416 ~]$
```

```
[user@host-766416 ~]$ kubectl get nodes -o yaml | grep -i schedulingdisabled
[user@host-766416 ~]$
```

**Comment by foxish** [comment]
That looks odd. The master doesn't have the `SchedulingDisabled` label either. I think this is a kubeadm bootstrapping issue. This looks related: https://github.com/kubernetes/kubeadm/issues/140. Some more context is in https://github.com/kubernetes/kubernetes/issues/36272. I'll have a longer discussion about this, but the annotation is about to turn into an API field in 1.6. 
 
Also, this is typical output:

```
NAME                           STATUS                     AGE
kubernetes-master              Ready,SchedulingDisabled   20d
kubernetes-minion-group-3pvp   Ready                      20d
kubernetes-minion-group-cnk6   Ready                      20d
kubernetes-minion-group-j0qh   Ready                      20d
```

In any case, @mccheah, I recommend just filtering on `unschedulable=true` for now, and leaving an issue open. I'll figure out the taints/tolerations and send a PR adding that sometime next week if it's necessary.

**Comment by mccheah** [comment]
Cool, I filed #73 to track further investigation.

One remark is that the `MultiServerFeignTarget` will make it such that a failed application submission will trigger another one to a different node. This could result in us pushing the jars and files over the wire multiple times, which could be expensive considering the size of the payload. I'm not too sure how significant this will be in practice however, especially since we are sending the data to different nodes every time.

**Comment by mccheah** [comment]
@foxish @aash Anything else to add here?

**Comment by ash211** [comment]
Nope.  Looks like we might remove the pinging process entirely in https://github.com/apache-spark-on-k8s/spark/pull/75 and might more proactively prune out nodes in the `MultiServerFeignTarget` in #73 but both of those are extensions to this.

Merging

**Pull request merged**